[GraphOptimizer] Fix the Conv-BatchNorm optimization

jfix71 opened the pull request:
*Description*: On Resnet50, many BatchNormalizationNodes were no longer being merged with the Convolution node before it. This commit fixes the problem.

It occurred because [this commit](https://github.com/pytorch/glow/commit/bd18db777cb775672ab47d626c019e979ac9edc8) removed trainable from Variables. This means that constant deduplication, which did not apply to trainable variables, now does apply and is merging 7 biases which are the same. Instead of giving up when the filter/bias have multiple uses, instead duplicate the filter/bias and allow the optimization to proceed.

*Testing*: Added a unit test

*Documentation*: N/A

qcolombet commented:
> However the optimization doesn't change the Conv's filter/bias, it just reads from them. So it's fine for the optimization to proceed.

Hmm, I am confused. Don't we modified them at line 974 and 988 respectively?

jfix71 commented:
Whoops! Yeah, I totally misread it 🙁 . We need to duplicate the constant. I'll update the PR and messages.

jfix71 commented:
Added a new version that duplicates Constants if necessary. This uncovered a problem in CSE for nodes with predicates, which I added a fix for in https://github.com/pytorch/glow/pull/1861. I will land this once that lands.

nadavrot commented:
Nice! Good catch!

qcolombet commented on lib/Optimizer/GraphOptimizer.cpp:
```diff
@@ -885,34 +885,52 @@ static void optimizePool(Function *F) {
 
+  // Create a new convolution node with the same parameters and input as CV, but
+  // with unique filter/bias which will be normalized below.
+  auto *newCV = F->createConv(CV.getName(), CV.getInput(), filterC, cbiasC,
```
Maybe add a comment that we need to create a new node because the constants may not be bound to that node.

Alternatively, we could always call setNThInput on the related arguments. BTW it may be better because it would avoid to rely on DCE to get rid of the now unconnected convolution node.

qcolombet commented:
BTW we might want to update the description of the PR in case we want to refer back to it.

qcolombet commented on lib/Optimizer/GraphOptimizer.cpp:
```diff
@@ -885,34 +885,58 @@ static void optimizePool(Function *F) {
+/// has a single use, \p node is returned. If \node is a Constant which has
+/// multiple uses, then \returns a new duplicate Constant that has the same
+/// contents as \p node.
+static Constant *getUniquelyUsedConstant(Function *F, Node &node) {
```
Nit: given we just use F->getParent(), maybe we should pass the module instead.

jfix71 replied:
Done

jfix71 commented:
BTW, this optimization also applies to https://github.com/pytorch/glow/issues/1641, as it is possible (albeit unlikely) a backend would prefer to not duplicate the weights/bias Constants and instead just perform the batch norm.
